stackslib/src/chainstate/burn/atc.rs

```diff
@@ -0,0 +1,1403 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use stacks_common::util::uint::Uint256;
+
+use crate::stacks_common::util::uint::BitArray;
+
+/// A fixed-point numerical representation for ATC.  The integer and fractional parts are both 64
+/// bits.  Internally, this is a Uint256 so that safe addition and multiplication can be done.
+///
+/// Bits 0-63 are the fraction.
+/// Bits 64-127 are the integer.
+/// Bits 128-256 are 0's to facilitate safe addition and multiplication.
+///
+/// The reasons we use this instead of f64 for ATC calculations are as follows:
+/// * This avoids unrepresentable states, like NaN or +/- INF
+/// * This avoids ambiguous states, like +0.0 and -0.0.
+/// * This integrates better into the sortition-sampling system, which uses a u256 to represent a
+/// probability range (which is what this is going to be used for)
+#[derive(Debug, Clone, PartialEq, Copy, Eq, Hash)]
+pub(crate) struct AtcRational(pub(crate) Uint256);
```

Comment:
Can you make the internal field private?

Reply:
Sure